kcidb/monitor/spool/__init__.py

```diff
@@ -97,29 +96,48 @@ def post(self, notification, timestamp=None):
             timestamp:      An "aware" datetime.datetime object specifying the
                             notification creation time, or None to use
                             datetime.datetime.now(datetime.timezone.utc).
+            due:            An "aware" datetime.datetime object specifying the
+                            time notification should be sent out,
+                            or None to use "timestamp".
 
         Returns:
             True, if the notification was posted onto the spool,
+            or an existing notification with timestamp past the
+            picked_until time was updated,
             False, if not (it was already there).
         """
         assert isinstance(notification, Notification)
         assert timestamp is None or \
             isinstance(timestamp, datetime.datetime) and timestamp.tzinfo
         if timestamp is None:
             timestamp = datetime.datetime.now(datetime.timezone.utc)
+        assert due is None or \
+            isinstance(due, datetime.datetime) and due.tzinfo
+        if due is None:
+            due = timestamp
+        message = notification.render().as_string(policy=email.policy.SMTPUTF8)
+        doc = self._get_doc(notification.id)
 
-        try:
-            self._get_doc(notification.id).create(dict(
+        @firestore.transactional
+        def create_or_update(transaction, message, doc):
+            snapshot = doc.get(field_paths=["picked_until"],
+                               transaction=transaction)
+            picked_until = snapshot.get("picked_until")
+            # If the document doesn't exist, or
+            # the timestamp is passed the picked_until field
+            if not picked_until or \
+                    timestamp > picked_until:
```

Comment:
So, how does this work if the document doesn't exist? Is `snapshot` returned in any case, and then it returns `None` for `"picked_until"`, or does `doc.get()` return `None` when the document doesn't exist? I just want to make sure we're handling both cases correctly :grin:

Reply:
BTW, could you not wrap this line? I think this will fit in a single line just fine.

Reply:
Hmm, yes doc.get() will return None when the document doesn't doesn't.
I am not sure I am handling that properly  
There is a method called exists to check if a doc exist, I think that will be better to handle an empty doc
from here: https://firebase.google.com/docs/firestore/query-data/get-data

Reply:
I think if it's documented that it will return None, it's enough to check for that. We need to get the `picked_until` field anyway, so better do it with just one call to the API (and to the "cloud").